rollup/internal/controller/relayer/l2_relayer_sanity.go

```diff
@@ -0,0 +1,438 @@
+			return fmt.Errorf("batch %d chunk %d has different codec version %d, expected %d", batch.Batch.Index, chunk.Index, chunk.CodecVersion, batch.Batch.CodecVersion)
+		}
+	}
+
+	// Validate each chunk individually
+	currentPrevChunk := prevChunk
+	for j, chunk := range batch.Chunks {
+		if err := r.validateSingleChunkConsistency(chunk, currentPrevChunk); err != nil {
+			return fmt.Errorf("batch %d chunk %d: %w", batch.Batch.Index, j, err)
+		}
+		currentPrevChunk = chunk
+	}
+
+	return nil
+}
+
+// validateSingleChunkConsistency validates a single chunk
+func (r *Layer2Relayer) validateSingleChunkConsistency(chunk *orm.Chunk, prevChunk *orm.Chunk) error {
+	if chunk == nil {
+		return fmt.Errorf("chunk is nil")
+	}
+
+	chunkHash := common.HexToHash(chunk.Hash)
+	if chunkHash == (common.Hash{}) {
+		return fmt.Errorf("chunk %d hash is zero", chunk.Index)
+	}
+
+	// Check chunk index continuity
+	if chunk.Index != prevChunk.Index+1 {
+		return fmt.Errorf("chunk index is not sequential: prev chunk index %d, current chunk index %d", prevChunk.Index, chunk.Index)
+	}
+
+	// Validate block range
+	if chunk.StartBlockNumber == 0 && chunk.EndBlockNumber == 0 {
+		return fmt.Errorf("chunk %d has zero block range", chunk.Index)
+	}
+
+	if chunk.StartBlockNumber > chunk.EndBlockNumber {
+		return fmt.Errorf("chunk %d has invalid block range: start %d > end %d", chunk.Index, chunk.StartBlockNumber, chunk.EndBlockNumber)
+	}
+
+	// Check hash fields
+	startBlockHash := common.HexToHash(chunk.StartBlockHash)
+	if startBlockHash == (common.Hash{}) {
+		return fmt.Errorf("chunk %d start block hash is zero", chunk.Index)
+	}
+
+	endBlockHash := common.HexToHash(chunk.EndBlockHash)
+	if endBlockHash == (common.Hash{}) {
+		return fmt.Errorf("chunk %d end block hash is zero", chunk.Index)
+	}
+
+	// Check block continuity with previous chunk
+	if prevChunk.EndBlockNumber+1 != chunk.StartBlockNumber {
+		return fmt.Errorf("chunk is not continuous with previous chunk %d: prev end block %d, current start block %d", prevChunk.Index, prevChunk.EndBlockNumber, chunk.StartBlockNumber)
+	}
+
+	// Check L1 messages continuity
+	expectedPoppedBefore := prevChunk.TotalL1MessagesPoppedBefore + prevChunk.TotalL1MessagesPoppedInChunk
+	if chunk.TotalL1MessagesPoppedBefore != expectedPoppedBefore {
+		return fmt.Errorf("L1 messages popped before is incorrect: expected %d, got %d", expectedPoppedBefore, chunk.TotalL1MessagesPoppedBefore)
+	}
+
+	return nil
+}
+
+// validateBlobsAgainstDatabase validates blobs against database records
+func (r *Layer2Relayer) validateBlobsAgainstDatabase(blobs []*kzg4844.Blob, batchesToValidate []*dbBatchWithChunks) error {
+	// Get codec for blob decoding
+	firstBatch := batchesToValidate[0].Batch
+	codec, err := encoding.CodecFromVersion(encoding.CodecVersion(firstBatch.CodecVersion))
+	if err != nil {
+		return fmt.Errorf("failed to get codec: %w", err)
+	}
+
+	// Validate each blob against its corresponding batch
+	for i, blob := range blobs {
```

Comment:
Not sure if this would be the right place but I think it would be good to also compare the batch hashes of the calldata + blob hash vs database